provider/pkg/provider/remote/copyController.go

```diff
@@ -0,0 +1,195 @@
+		failures = append(failures, p.CheckFailure{
+			Property: "archive",
+			Reason:   "archive must be a path to a file or directory",
+		})
+	}
+
+	return inputs, failures, nil
+}
+
+func doCopy(ctx p.Context, input CopyInputs) (CopyOutputs, error) {
+	sourcePath := input.sourcePath()
+
+	ctx.Logf(diag.Debug,
+		"Creating remote: %s:%s from local '%s'",
+		input.Connection.Host, input.RemotePath, sourcePath)
+
+	client, err := input.Connection.Dial(ctx)
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	defer client.Close()
+
+	sftp, err := sftp.NewClient(client)
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	defer sftp.Close()
+
+	src, err := os.Open(sourcePath)
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	defer src.Close()
+
+	srcInfo, err := src.Stat()
+	if err != nil {
+		return CopyOutputs{input}, err
+	}
+	if srcInfo.IsDir() {
+		err = copyDir(sftp, sourcePath, input.RemotePath)
+	} else {
+		err = copyFile(sftp, sourcePath, input.RemotePath)
+	}
+	return CopyOutputs{input}, err
+}
+
+// This is the Create method. This will be run on every Copy resource creation.
+func (*Copy) Create(ctx p.Context, name string, input CopyInputs, preview bool) (string, CopyOutputs, error) {
+	if preview {
+		return "", CopyOutputs{input}, nil
+	}
+
+	outputs, err := doCopy(ctx, input)
+	if err != nil {
+		return "", CopyOutputs{input}, err
+	}
+
+	id, err := resource.NewUniqueHex("", 8, 0)
+	return id, outputs, err
+}
+
+func (c *Copy) Update(ctx p.Context, id string, olds CopyOutputs, news CopyInputs, preview bool) (CopyOutputs, error) {
+	needCopy := true
+	if news.hash() == olds.hash() && news.RemotePath == olds.RemotePath {
+		needCopy = false
+	}
+
+	if preview {
+		// TODO,tkappler how to show the user whether we need to copy or not?
+		return CopyOutputs{news}, nil
+	}
+
+	if needCopy {
+		return doCopy(ctx, news)
+	}
+	return CopyOutputs{news}, nil
+}
+
+func copyFile(sftp *sftp.Client, src, dst string) error {
+	local, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer local.Close()
+
+	remote, err := sftp.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer remote.Close()
+
+	_, err = remote.ReadFrom(local)
+	return err
+}
+
+// copyDir copies a directory recursively from the local file system to a remote host.
+// Note that the current is naive and sequential and therefore can be slow.
+func copyDir(sftp *sftp.Client, src, dst string) error {
+	fileSystem := os.DirFS(src)
+	return fs.WalkDir(fileSystem, ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		remotePath := filepath.Join(dst, path)
+
+		if !d.IsDir() {
+			return copyFile(sftp, filepath.Join(src, path), remotePath)
+		}
+
+		dirInfo, err := sftp.Stat(remotePath)
+		// sftp normalizes the error to os.ErrNotExist, see client.go: normaliseError.
+		if err != nil && !errors.Is(err, os.ErrNotExist) {
+			return err
+		}
+
+		if dirInfo == nil {
+			return sftp.Mkdir(remotePath)
```

Comment:
What if the parent of `dst` doesn't exist? This would end up failing, right? Should we maybe use `MkdirAll` here to be safe? Or maybe better do that in the `doCopy` function because `copyFile` needs it to, right?

Reply:
In a recursive call, that shouldn't be possible.

In the first call - creating the user's top-level directory - one could argue that it should fail because the user asked to copy to a non-existent location. It's one of the questions in my (internal) ideation doc.